dynawo/sources/Models/Modelica/Dynawo/NonElectrical/Blocks/Continuous/TransferFunction.mo

```diff
@@ -16,46 +16,38 @@ within Dynawo.NonElectrical.Blocks.Continuous;
 block TransferFunction "Linear transfer function"
   extends Modelica.Blocks.Interfaces.SISO(y(start = y_start));
 
-  parameter Real b[:] = {1}
-    "Numerator coefficients of transfer function (e.g., 2*s+3 is specified as {2,3})";
-  parameter Real a[:] = {1}
-    "Denominator coefficients of transfer function (e.g., 5*s+6 is specified as {5,6})";
+  parameter Real b[:] = {1} "Numerator coefficients of transfer function (e.g., 2*s+3 is specified as {2,3})";
+  parameter Real a[:] = {1} "Denominator coefficients of transfer function (e.g., 5*s+6 is specified as {5,6})";
+  parameter Boolean ZeroA1 = false "If true, highest-order coefficient of denominator can be zero";
 
-  output Real x[nx](start = x_start)
-    "State of transfer function from controller canonical form";
+  output Real x[nx](start = x_start) "State of transfer function from controller canonical form";
 
-  parameter Real x_start[nx] = zeros(nx)
-    "Initial or guess values of states"
-    annotation(Dialog(group = "Initialization"));
-  parameter Real y_start = 0
-    "Initial value of output (derivatives of y are zero up to nx-1-th derivative)"
-    annotation(Dialog(group = "Initialization"));
+  parameter Real x_start[nx] = zeros(nx) "Initial or guess values of states" annotation(
+    Dialog(group = "Initialization"));
+  parameter Real y_start = 0 "Initial value of output (derivatives of y are zero up to nx-1-th derivative)" annotation(
+    Dialog(group = "Initialization"));
 
 protected
   parameter Integer na = size(a, 1) "Size of denominator of transfer function";
   parameter Integer nb = size(b, 1) "Size of numerator of transfer function";
-  parameter Integer nx = size(a, 1) - 1;
-  parameter Real bb[:] = vector([zeros(max(0,na-nb),1);b]);
-  parameter Real d = bb[1] / a_one;
-  parameter Real a_one = if a[1] > 100 * Modelica.Constants.eps * sqrt(a*a) then a[1] else 1;
-  parameter Real a_end = if a[end] > 100 * Modelica.Constants.eps * sqrt(a*a) then a[end] else 1;
+  parameter Integer nx = size(a, 1) - 1 "Size of vector x";
+  parameter Real bb[:] = vector([zeros(max(0,na-nb),1);b]) "Numerator coefficients, padded if necessary by zeroes for highest-order coefficients";
+  parameter Real d = bb[1] / a_one "Ratio of highest-order coefficients of numerator and denominator";
+  parameter Real a_one = if a[1] > 100 * Modelica.Constants.eps then a[1] else 1 "Non-zero value of highest-order coefficient of denominator";
+  parameter Real a_end = if a[end] > 100 * Modelica.Constants.eps * sqrt(a*a) then a[end] else 1 "Non-zero value of lowest-order coefficient of denominator";
   Real x_scaled[nx](start = a_end * x_start) "Scaled vector x";
 
 equation
   assert(size(b,1) <= size(a,1), "Transfer function is not proper");
 
-  if a[1] <= 100 * Modelica.Constants.eps then
+  if ZeroA1 and a[1] <= 100 * Modelica.Constants.eps then
```

Comment:
So here I still think that the condition should only be  if a[1] <= 100 * Modelica.Constants.eps, wouldn't be enough?

Reply:
It would work, but if ZeroA1 = false, this if branch is ignored, thus accelerating the calculation. Without the parameter ZeroA1, the non-regression tests fail.